xwiki-platform-core/xwiki-platform-annotation/xwiki-platform-annotation-ui/src/main/resources/AnnotationCode/Script.xml

```diff
@@ -976,28 +977,193 @@ XWiki.Annotation = Class.create({
+  */
+  markAnnotation: function(markedNode, ann) {
+    var wrapper = document.createElement('span');
+    wrapper.addClassName('annotation');
+    wrapper.addClassName('ID' + ann.annotationId);
+
+    var parentNode = markedNode.parentElement;
+    parentNode.replaceChild(wrapper, markedNode);
+    wrapper.appendChild(markedNode);
+  },
+
+  /**
+   * For the first and last node, split the nodes at the known offset to not annotate the whole text. Create a new range
+   * with these new nodes.
+   *
+   * @param range the DOM range of the annotated text
+  */
+  fixRangeEnds: function(range) {
+    var strictRange = new Range();
+
+    var newStartNode = range.startContainer.splitText(range.startOffset);
+    strictRange.setStart(newStartNode, 0);
+
+    var newEndNode = range.endContainer.splitText(range.endOffset);
+    strictRange.setEnd(newEndNode, 0);
+
+    return strictRange;
+  },
+
+  /**
+   * Create a DOM Range by knowing the start and end index from inside the plain content of the element.
+   *
+   * @param annotatedElement the element from where the range is constructed
+   * @param startIndex start offset where the range begins
+   * @param endIndex end offset where the range end
+  */
+  getDOMRange: function(annotatedElement, startIndex, endIndex) {
+    this.annotatedElementPlainText = "";
+    var startPosition = this.getElementAtOffset(annotatedElement, startIndex);
+    this.annotatedElementPlainText = "";
+    var endPosition = this.getElementAtOffset(annotatedElement, endIndex);
+    var range = new Range();
+    range.setStart(startPosition.element, startPosition.offset);
+    range.setEnd(endPosition.element, endPosition.offset);
+    return range;
+  },
+
+  /**
+   * Knowing the offset relative to the full plain content of the root element, get the corresponding child node that contains it and the offset
+   * specific to the new node. The DOM is traversed recursively starting with the root node and the plain content is computed to know when the
+   * wanted element is found.
+   *
+   * @param rootElement the element where the search is done
+   * @param offset the offset relative to the full plain content
+  */
+  getElementAtOffset: function(rootElement, offset) {
+    var childNodes = rootElement.childNodes;
+    var child;
+    for (let i = 0; i &lt; childNodes.length; i++) {
+      child = childNodes[i];
+      if (child &amp;&amp; child.nodeType == 3) {
+        var initialLength = this.annotatedElementPlainText.length;
+        this.annotatedElementPlainText += child.textContent.replace(/\s/g, '');
+        if (offset &lt; this.annotatedElementPlainText.length) {
+          var elementSpecificOffset = this.getElementSpecificOffset(child, offset - initialLength);
+          return {"element": child, 'offset': elementSpecificOffset};
+        }
+      }
+      if (child.childNodes.length &gt; 0) {
+        var maybeFoundElement = this.getElementAtOffset(child, offset);
```

Comment:
Here you need to remove from the offset the current length of plain text (within the current parent node).